src/main/scala/chiseltest/internal/ThreadedBackend.scala

```diff
@@ -36,14 +35,183 @@ case class ForkBuilder(name: Option[String], region: Option[Region], threads: Se
+    throw new ClockResolutionException(
+      "ICR not available on chisel-testers2 / firrtl master"
+    )
+  }
+
+  def run: AnnotationSeq = Context.context.withValue(Some(this)) {
+    rootTimescope = Some(new RootTimescope)
+    val mainThread = new TesterThread(
+      () => {
+
+        /** @todo make user be able to set his with his own init logic,
+          *       for example reset, local last state or etc.
+          */
+        simulatorInterface.poke("reset", 1)
+        simulatorInterface.step(1)
+        simulatorInterface.poke("reset", 0)
+        testFunction(dut)
+      },
+      TimeRegion(0, Region.default),
+      rootTimescope.get,
+      0,
+      Region.default,
+      None
+    )
+    mainThread.thread.start()
+    require(allThreads.isEmpty)
+    allThreads += mainThread
+
+    try {
+      while (!mainThread.done) { // iterate timesteps
+        clockCounter.put(dut.clock, getClockCycle(dut.clock) + 1)
+
+        logger.debug(s"clock step")
+
+        /** @todo allow dependent clocks to step based on test stimulus generator
+          *       remove multiple invocations of getClock
+          */
+        // Unblock threads waiting on dependent clock
+        val steppedClocks = Seq(dut.clock) ++ lastClockValue.collect {
+          case (clock, lastValue) if getClock(clock) != lastValue && getClock(clock) =>
+            clock
+        }
+
+        /** @todo ignores cycles before a clock was stepped on. */
+        steppedClocks.foreach { clock =>
+          clockCounter.put(dut.clock, getClockCycle(clock) + 1)
+        }
+        lastClockValue.foreach {
+          case (clock, _) =>
+            lastClockValue.put(clock, getClock(clock))
+        }
+
+        runThreads(steppedClocks.toSet)
+
+        idleLimits.foreach {
+          case (clock, limit) =>
+            idleCycles.put(clock, idleCycles.getOrElse(clock, -1) + 1)
+            if (idleCycles(clock) >= limit) {
+              chiselTestExceptions.append(new TimeoutException(s"timeout on $clock at $limit idle cycles"))
+            }
+        }
 
-  override def setTimeout(signal: Clock, cycles: Int): Unit = {
+        simulatorInterface.step(1)
+      }
+      annotations
+    } finally {
+      rootTimescope = None
+
+      for (thread <- allThreads.clone()) {
+        // Kill the threads using an InterruptedException
+        if (thread.thread.isAlive) {
+          thread.thread.interrupt()
+        }
+      }
+
+      simulatorInterface.finish() // Do this to close down the communication
+    }
+  }
+
+  /** Sets the value associated with a key in a per-test map.
+    */
+  def setVar(key: Any, value: Any): Unit = {
+    testMap.put(key, value)
+  }
+
+  /** Returns the value associated with the key in a per-test map.
+    */
+  def getVar(key: Any): Option[Any] = {
+    testMap.get(key)
+  }
+
+  /** Check a signal is equal to a value,
+    * if not equal, store this exception to [[chiselTestExceptions]].
+    */
+  def expectBits(
```

Comment:
I think expectBits belongs in the simulator interface. Note that in the current design, BackendInterface is meant to be the simulator interface, so that backends could implement optimized expects instead of needing to roundtrip to Scala each time. Having expect as a "primitive" operation also allows things like SteppedHWIOTester which generates test using a certain subset of primitives down to hardware. I'm not convinced we want to architecturally foreclose that route, even if we haven't used it yet.

Reply:
I think so. but in current behavior, `expectBits` is implemented in Scala to directly compare.
I also think in the future. After Verification API landing to Chisel. I think `expect` API can finally deprecated. let ScalaTest only `peek`/`poke`, let `assume` `assert` `cover` stuffs to simulator, and expose correspond error to chisletest.
but in current implementation, maybe we just can leave this as it is?